src/audio/copier/copier_gain.c

```diff
@@ -79,3 +79,79 @@ enum copier_gain_state copier_gain_eval_state(struct copier_gain_params *gain_pa
 
 	return state;
 }
+
+int copier_gain_dma_control(uint32_t node_id, const uint32_t *config_data,
+			    size_t config_size, enum sof_ipc_dai_type dai_type)
+{
+	union ipc4_connector_node_id node = (union ipc4_connector_node_id)node_id;
+	struct gain_dma_control_data *gain_data = NULL;
+	struct ipc_comp_dev *icd = NULL;
+	struct processing_module *mod = NULL;
+	struct copier_data *cd = NULL;
+	struct ipc *ipc = ipc_get();
+	struct list_item *clist;
+	struct comp_dev *dev;
+	int ret;
+
+	list_for_item(clist, &ipc->comp_list) {
+		icd = container_of(clist, struct ipc_comp_dev, list);
+
+		if (icd->type != COMP_TYPE_COMPONENT)
+			continue;
+
+		dev = icd->cd;
+
+		if (dev->ipc_config.type != SOF_COMP_DAI)
+			continue;
+
+		mod = comp_mod(dev);
+		cd = module_get_private_data(mod);
+
+		switch (dai_type) {
+		default:
+			comp_warn(dev, "Gain DMA control: no dai type=%d found", dai_type);
+			break;
+		}
```

Comment:
that will also be an always-true warning.

Not sure where this programming pattern comes from but it'd be frowned upon by all reasonable static checkers.